### PR TITLE
[dynamo] Restrict support for `out=` variants of torch operators

### PR DESCRIPTION
**Pull request by StrongerXi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #140155
* #140154
* #140153
* #140435
* #140436
* #140152
* #140330
* __->__ #140202
* #140201
* #140151
* #140150
* #140149
* #140036
* #140035

There has been a series of attempts to provide support for resizing in
torch operators like `torch.sigmoid(x, out=y)`, i.e., `y` would have a
different shape before and after this expression. Prior to this patch,
we have some checks to graph break if the shape changed.

This patch extends
1. extends the existing check and graph break for any shape change, not
   just for `TensorVariable` with source field.
2. removes an old code path which was introduced to address the shape
   change, but became obselete in that regard because we added extra
   checks to graph break upon shape change. Moreover, this old code path
   is unsound, it tries to replace references to the old
   `TensorVariable` the new one returned by `wrap_fx_proxy`, but it only
   does the replacement in `symbolic_locals`, which breaks when cells
   are involved. In general the old `TensorVariable` could be _anywhere_,
   think the `replace_all` we had for immutable VTs.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

**Comment by vadimkantorov** [comment]
I also wonder, can Inductor generate inplace variants? (e.g. one could imagine Inductor to generated inplae code for fused ReLU+Dropout by itself https://github.com/pytorch/pytorch/issues/92927)

**Review by jansel — APPROVED** [review]

**Comment by StrongerXi** [comment]
Rebase

**Comment by StrongerXi** [comment]
Rebase

**Comment by StrongerXi** [comment]
> I also wonder, can Inductor generate inplace variants? (e.g. one could imagine Inductor to generated inplae code for fused ReLU+Dropout by itself #92927, e.g. for `torch.compile(nn.Sequential(nn.ReLU(inplace=True), nn.Dropout(inplace=True)))`)

I don't have much context there, but it doesn't feel relevant to the change in this patch..?
